File: 01-Languages/TypeScript/21-Solutions/solution-02.ts
// solution-02.ts - discriminated Shape union with exhaustiveness checking (Exercise 02)

interface Circle {
  kind: "circle";
  radius: number;
}
interface Rectangle {
  kind: "rectangle";
  width: number;
  height: number;
}
interface Triangle {
  kind: "triangle";
  base: number;
  height: number;
}
type Shape = Circle | Rectangle | Triangle;

function area(shape: Shape): number {
  switch (shape.kind) {
    case "circle":
      return Math.PI * shape.radius ** 2;
    case "rectangle":
      return shape.width * shape.height;
    case "triangle":
      return 0.5 * shape.base * shape.height;
    default: {
      // If a fourth Shape variant is ever added without a matching `case` above,
      // `shape` here stops being assignable to `never` and this line fails to compile --
      // that's the whole point of an exhaustiveness check: the bug is caught at build time.
      const exhaustive: never = shape;
      return exhaustive;
    }
  }
}

function describe(shape: Shape): string {
  return `${shape.kind}: area = ${area(shape).toFixed(2)}`;
}

const shapes: Shape[] = [
  { kind: "circle", radius: 5 },
  { kind: "rectangle", width: 4, height: 6 },
  { kind: "triangle", base: 3, height: 8 },
];

for (const shape of shapes) {
  console.log(describe(shape));
}

File: 01-Languages/CSharp/21-Solutions/solution-02.cs
// solution-02.cs - Pattern Matching Shape Calculator
// See: ../20-Exercises/README.md#exercise-02--pattern-matching-shape-calculator-beginnerintermediate
//
// Run with:
//     dotnet run solution-02.cs

Console.WriteLine("--- switch expression with type + property patterns ---");
Console.WriteLine($"Circle(3): {Area(new Circle(3)):F2}");
Console.WriteLine($"Rectangle(4,5): {Area(new Rectangle(4, 5)):F2}");
Console.WriteLine($"Rectangle(4,4) (square guard): {Area(new Rectangle(4, 4)):F2}");
Console.WriteLine($"Triangle(6,2): {Area(new Triangle(6, 2)):F2}");

try {
    Area("not a shape");
} catch (UnknownShapeException ex) {
    Console.WriteLine($"Caught expected error: {ex.Message}");
}

static double Area(object shape) => shape switch {
    Circle(var r) => Math.PI * r * r,

    // Property pattern with a 'when' guard: a Rectangle whose Width equals
    // its Height is still a Rectangle at the type level, but worth calling
    // out distinctly before falling through to the general rectangle case.
    Rectangle { Width: var w, Height: var h } when w == h =>
        LogSquareAndReturnArea(w, h),

    Rectangle(var w, var h) => w * h,
    Triangle(var b, var h) => 0.5 * b * h,

    // Discard pattern: anything that isn't one of the shapes above is a
    // programming error, not a recoverable case, so it throws.
    _ => throw new UnknownShapeException($"Unrecognized shape: {shape}")
};

static double LogSquareAndReturnArea(double w, double h) {
    Console.WriteLine($"  (detected a square: {w}x{h})");
    return w * h;
}

record Circle(double Radius);
record Rectangle(double Width, double Height);
record Triangle(double Base, double Height);

class UnknownShapeException : Exception {
    public UnknownShapeException(string message) : base(message) {}
}
